editor/src/clj/editor/slice9.clj

```diff
@@ -0,0 +1,241 @@
+;; Copyright 2020-2022 The Defold Foundation
+;; Copyright 2014-2020 King
+;; Copyright 2009-2014 Ragnar Svensson, Christian Murray
+;; Licensed under the Defold License version 1.0 (the "License"); you may not use
+;; this file except in compliance with the License.
+;;
+;; You may obtain a copy of the License, together with FAQs at
+;; https://www.defold.com/license
+;;
+;; Unless required by applicable law or agreed to in writing, software distributed
+;; under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+;; CONDITIONS OF ANY KIND, either express or implied. See the License for the
+;; specific language governing permissions and limitations under the License.
+
+(ns editor.slice9
+  (:require [editor.geom :as geom]))
+
+(set! *warn-on-reflection* true)
+
+(defn- box-corner-coords->vertices2 [[x0 y0 x1 y1]]
+  [[x0 y0] [x0 y1] [x1 y1] [x1 y0]])
+
+(defn- rotated-box-corner-coords->vertices2 [[x0 y0 x1 y1]]
+  [[x0 y0] [x1 y0] [x1 y1] [x0 y1]])
+
+(defn- box-corner-coords->vertices3 [[x0 y0 x1 y1]]
+  [[x0 y0 0.0] [x0 y1 0.0] [x1 y1 0.0] [x1 y0 0.0]])
+
+(defn- box-corner-coords->vertices4 [[x0 y0 x1 y1]]
+  [[x0 y0 0.0 1.0] [x0 y1 0.0 1.0] [x1 y1 0.0 1.0] [x1 y0 0.0 1.0]])
+
+(defn- ranges->box-corner-coords [x-ranges y-ranges]
+  (for [[x0 x1] x-ranges
+        [y0 y1] y-ranges]
+    [x0 y0 x1 y1]))
+
+(defn- ranges->rotated-box-corner-coords [x-ranges y-ranges]
+  (for [[y0 y1] y-ranges
+        [x0 x1] x-ranges]
+    [x0 y0 x1 y1]))
+
+(defn- steps->ranges [v]
+  (partition 2 1 v))
+
+(defn- pivot->h-align [pivot]
+  (case pivot
+    (:pivot-e :pivot-ne :pivot-se) :right
+    (:pivot-center :pivot-n :pivot-s) :center
+    (:pivot-w :pivot-nw :pivot-sw) :left))
+
+(defn- pivot->v-align [pivot]
+  (case pivot
+    (:pivot-ne :pivot-n :pivot-nw) :top
+    (:pivot-e :pivot-center :pivot-w) :middle
+    (:pivot-se :pivot-s :pivot-sw) :bottom))
+
+(defn- pivot-offset [pivot size]
+  (let [h-align (pivot->h-align pivot)
+        v-align (pivot->v-align pivot)
+        xs (case h-align
+             :right -1.0
+             :center -0.5
+             :left 0.0)
+        ys (case v-align
+             :top -1.0
+             :middle -0.5
+             :bottom 0.0)]
+    (mapv * size [xs ys 1])))
+
+(def ^:private box-triangles-vertex-order [0 1 3 3 1 2])
+
+(defn box->triangle-vertices [box]
+  ;; box vertices are in order BL TL TR BR
+  ;;
+  ;; turns:
+  ;; 1---2
+  ;; |   |
+  ;; 0---3
+  ;;
+  ;; into:
+  ;; 1     1-2
+  ;; |\  +  \|
+  ;; 0-3     3
+  ;;
+  ;; normal uv box vertices are in order BL TL TR BR
+  ;;
+  ;; turns:
+  ;; 1---2
+  ;; |   |
+  ;; 0---3
+  ;;
+  ;; into:
+  ;; 1     1-2
+  ;; |\  +  \|
+  ;; 0-3     3
+  ;;
+  ;; _rotated_ uv box vertices are in order TL TR BR BL
+  ;;
+  ;; turns:
+  ;;
+  ;; 0---1
+  ;; |   |
+  ;; 3---2
+  ;;
+  ;; into:
+  ;; 0-1     1
+  ;; |/  +  /|
+  ;; 3     3-2
+  (map box box-triangles-vertex-order))
+
+(defn vertex-data
```

Comment:
This function is more or less a 100% copy-paste as well, except for the return values. Maybe we can merge this as well now, but I didn't feel confident enough on how to do that the correct way :'(